### ISSUE TITLE
Elastic Container Service: Waiting for deployment

### ISSUE THREAD

**Issue by dfonnegra** [aws/aws-cdk#14994]
## :question: General Issue

I built a stack for my current project using ECS and ALB. I had deployed the development environment and it was working perfectly. The problem came when I tried to deploy for the first time the staging environment. Well, it happens that the stack deployment waits for the ECS service to be deployed! But as it is the first deployment the ECR containers don't have images yet. These images are created when the CodePipeline is run, but the CodePipeline doesn't exist yet because it has a deployment stage that depends on the ECS service, so the AWS CDK hasn't deployed it yet.

So to summarize, the ECS service doesn't deploy because it depends on the CodePipeline but the CodePipeline doesn't deploy because it depends on the ECS service.

### The Question
Is there a clean way to solve this? I thought of creating an independent CodeBuild project that creates the images, but I find it ugly because it's code duplication.

### Other information
I saw that a guy reported something similar some time ago and there was an interesting discussion about it [here](https://github.com/aws/aws-cdk/issues/5335) for implementing a fire-and-forget capability.
I still don't understand why the system is forcing you to deploy the code while deploying the infrastructure, they don't have anything to do with each other. If I want to deploy a cluster setup, it shouldn't matter if the code is correctly running or not, that should be a task of Code Deploy and the Load Balancer.

**Comment by ryparker** [comment]
Hey @dfonnegra :wave:

I'd like to offer some help.
Let me make sure I understand this right, bare with me...

```
We have a stack that consists of:
- ECS
- ALB
- CodePipeline

1. We deployed this stack as a development environment successfully.
2. We'd like to deploy that same stack as a staging environment.

Problem: When attempting to deploy the stack to a staging environment there's a dependency error because the ECS resource requires an image that is created by the CodePipeline resource.
```

Do I have that correct?

If so, how were you able to deploy the development stack without running into this issue? Are you using the same region for both deployments?

Any example code of your setup would be helpful as well. 😄

**Comment by dfonnegra** [comment]
Yes. On staging, it deployed successfully because I was deploying the stack gradually as soon as I was writing the code. So when I was creating the CodePipeline, there was already an ECR image and an ECS cluster.

To summarize a little bit the issue, I have the following setup:
- ECR
- ECS
    - One service
    - Two task definitions
- ALB
- CodePipeline
    - Source Github
    - Build Project that pushes the ECR image.
    - ECS Deploy.
    
So let's suppose that I've never deployed this stack. So it happens the following:
1. Deploys the ECR repos, which are empty at the beginning.
2. As the CodePipeline depends on the ECS service, it starts deploying the ECS service.
3. The ECS service won't be deployed until the health checker of the ALB says "this service is healthy".
4. It's not healthy because the ECR repos are empty.
5. I can't fill the repos because the CodePipeline hasn't been created.
6. Deadlock.

**Comment by ryparker** [comment]
One possible solution would be to separate this into separate stacks.

For example:

```markdown
# base-stack.ts
- Deploys ECR repos
- Deploys CodeBuild projects for building ECR images

# infrastructure-pipeline-stack.ts
- Deploys CodePipeline:
	1. Stage: Source
		- Source infra code from GitHub
	2. Stage: Test (optional: Always good practice to validate your changes will not break prod)
	3. Stage: Deploy-To-Prod
		- Deploys ALB
		- Deploys ECS
		- Deploy services/task definitions

# Deployment steps:
1. Deploy base-stack.ts
2. Commit code to service repos in order to trigger the CodeBuild projects 
   that we just deployed which will build and store the first ECR images.
3. Deploy infrastructure-pipeline-stack.ts
```

When a change is made to one of your CodeBuild project repos, it should trigger a new ECR image to be built and uploaded. At this point your new ECR images will not auto deploy to the ECS service. To deploy new images to your service I've seen it done a variety of ways but here a few:

- Manually via console: login to the console's ECS page and force deploy a new task to the service.
- Manually via CLI: `aws ecs update-service --cluster <cluster name> --service <service name> --force-new-deployment`
- Automatically: Create a new CodePipeline that looks like the following:

```markdown
# myWebService-pipeline-stack.ts
- Deploys CodePipeline:
	1. Stage: Source
	   - Source latest ECR image from ECR repo by using the CDK's `EcrSourceAction`
	2. Stage: Deploy
	    - Deploy latest ECR image to ECS service by using the CDK's `EcsDeployAction`

# Deployment note:
This would need to be deployed after step 3 from above ("Deploy infrastructure-pipeline-stack.ts") so that the ECS service is already available.
```

refs: [`EcrSourceAction`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codepipeline-actions.EcrSourceAction.html), [`EcsDeployAction`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codepipeline-actions.EcsDeployAction.html)


Sorry for the wait. Let me know if this is helpful 🙂

**Comment by ryparker** [comment]
One additional note to my post above: 
If you really wanted to keep all your infrastructure code in one stack then instead of having a `base-stack.ts` and `infrastructure-pipeline-stack.ts` you could do something like the following:


```markdown
# infrastructure-stack.ts
- Deploys ECR repos
- Deploys CodeBuild projects for building ECR images
- Deploys CodePipeline:
	1. Stage: Source
		- Source infra code from GitHub
	2. Stage: Upload-Images-To-ECR
	    - Source ECR project code from repo(s)
	    - Build image(s)
	    - Upload image(s) to ECR
	3. Stage: Test (optional: Always good practice to validate your changes will not break prod)
	4. Stage: Deploy-To-Prod
		- Deploys ALB
		- Deploys ECS
		- Deploy services/task definitions

# Deployment steps:
1. Deploy infrastructure-pipeline-stack.ts

Note: you would still need to setup another pipeline if you wanted ECR updates to auto deploy to 
your service
```

The reason I'd prefer to not have a stage for "Upload-Image-To-ECR" In your infra pipeline is because this is only necessary on your first deployment. Subsequent updates to your infrastructure may not require a new ECR image to be uploaded. Which adds unnecessary time and dependencies to your infrastructure pipeline.

**Comment by dfonnegra** [comment]
Yeah, as I commented before, the only solution I see for now is creating a separate Build project that deploys the ECR images. That solution for me breaks the whole point of the pipelines. They're supposed to be a system that forces the deployer, to do things in the right order. If I have a Build project and a Pipeline separated, the deployer could run first the pipeline and then the build, and the deployment wouldn't work because you deployed the images before they were updated. With that solution, I better create a Build project, a code deploy project, and create a manual that specifies the order of deployment. The concept of "pipeline" becomes useless.

**Comment by ryparker** [comment]
>  If I have a Build project and a Pipeline separated, the deployer could run first the pipeline and then the build, and the deployment wouldn't work because you deployed the images before they were updated.

I'm not sure what you mean. The pipeline I mentioned in my last comment would in fact upload the latest images to ECR before deploying your infrastructure & services.

> With that solution, I better create a Build project, a code deploy project, and create a manual that specifies the order of deployment. The concept of "pipeline" becomes useless.

What do you mean by "code deploy project" and "manual"? CodePipeline is designed to provide you with enough utility to control the order of deployments via stages and steps.

It may be helpful to provide some of the source code you're using so that I can better understand the scenario you're trying to tackle.

**Comment by dfonnegra** [comment]
Hi, I think I haven't thanked you so far for your time, but thanks I really appreciate it.

> Is there a clean way to solve this? I thought of creating an independent CodeBuild project that creates the images, but I find it ugly because it's code duplication.

So as mentioned here, I considered creating a CodeBuild project to deploy the images. This CodeBuild project will be run only once as the CodePipeline would push the ECR image after the first image creation. I find it unclean and you'll have a CodeBuild project that will become useless after the first build.

This might look like a small detail but as projects grow, having these small sneaky details for the deployments will make at some point the deployment process a pain in the ass.

**Comment by ryparker** [comment]
No problem. Although I think that there's some misunderstanding in how CodeBuild and CodePipeline can be used together.

> I considered creating a CodeBuild project to deploy the images. This CodeBuild project will be run only once because, once the CodePipeline is deployed, the CodePipeline will be in charge of pushing the ECR images. I find it unclean and you'll have a CodeBuild project that will become useless after the first build.

In this case CodeBuild is used to upload images to ECR. It can be ran independently or triggered from within a CodePipeline ([see CDK's CodeBuildAction](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codepipeline-actions.CodeBuildAction.html)). This way you can reuse the same CodeBuild project and keep your code DRY.

Take my `infrastructure-stack.ts` example from above:

```markdown
# infrastructure-stack.ts
- Deploys ECR repos
- Deploys CodeBuild projects for building ECR images
- Deploys CodePipeline:
	1. Stage: Source
		- Source infra code from GitHub
	2. Stage: Upload-Images-To-ECR
	    - Source ECR project code from repo(s)
	    - Build image(s) & Upload image(s) to ECR
	3. Stage: Test (optional: Always good practice to validate your changes will not break prod)
	4. Stage: Deploy-To-Prod
		- Deploys ALB
		- Deploys ECS
		- Deploy services/task definitions
```

In this example we `deploy CodeBuild projects for building ECR images` which are reused in the second stage `Upload-Images-To-ECR`.

In CDK code it might look something like this:

```ts
import { App, Stack } from '@aws-cdk/core';
import {
  Role,
  ServicePrincipal,
  PolicyDocument,
  PolicyStatement,
  Effect,
} from '@aws-cdk/aws-iam';
import { Pipeline, Artifact } from '@aws-cdk/aws-codepipeline';
import {
  CodeStarConnectionsSourceAction,
  CodeBuildAction,
} from '@aws-cdk/aws-codepipeline-actions';
import {
  PipelineProject,
  ComputeType,
  BuildSpec,
  LinuxBuildImage,
} from '@aws-cdk/aws-codebuild';

const app = new App();

const infrastructureStack = new Stack(app, 'infrastructure-stack')

// IAM Roles
const CodeBuildRole = new Role(infrastructureStack, 'CodeBuild-Role', {
  assumedBy: new ServicePrincipal('codebuild.amazonaws.com'),
  description: 'Role used for CodeBuild Projects',
  inlinePolicies: {
    CodeBuildNestedCFNAccessPolicy: new PolicyDocument({
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['*'],
          resources: ['*'],
        }),
      ],
    }),
  },
});

// CodeBuild Projects
const buildAndUploadEcrImageCodeBuildProject = new PipelineProject(infrastructureStack, 'build-and-upload-to-ecr-codebuild', {
  projectName: 'build-and-upload-to-ecr',
  role: CodeBuildRole,
  environment: {
    buildImage: LinuxBuildImage.STANDARD_5_0,
    computeType: ComputeType.SMALL,
  },
  // This buildspec handles building the image and then uploads the image to ECR
  buildSpec: BuildSpec.fromSourceFilename(
    'buildspec.yml'
  ),
})

// Artifacts
const imageSourceArtifact = new Artifact('imageSourceArtifact');
const imageBuildArtifact = new Artifact('imageBuildArtifact');

// Pipelines
const infrastructureCodePipeline = new Pipeline(infrastructureStack, 'Infrastructure-CodePipeline', {
  pipelineName: 'infrastructure',
  stages: [
    {
      stageName: 'Upload-Images-To-ECR',
      actions: [
        // Source ECR project code from repo(s)
        new CodeStarConnectionsSourceAction({
          actionName: 'GitHub',
          connectionArn: 'CODESTAR_ARN',
          owner: 'GITHUB_OWNER_NAME',
          repo: 'GITHUB_REPO_NAME',
          branch: 'main',
          output: imageSourceArtifact,
          runOrder: 1,
        }),
        // Build image(s) & Upload image(s) to ECR
        new CodeBuildAction({
          actionName: 'build-and-upload-images-to-ecr',
          input: imageSourceArtifact,
          project: buildAndUploadEcrImageCodeBuildProject, // <-- CodeBuild project is being used in pipeline
          outputs: [imageBuildArtifact],
          runOrder: 2,
        }),
      ],
    },
    // Other stages you wish to chain after ECR images are uploaded.
  ]
})
```
_The above code is for example purposes. You'll have to adjust this to best fit your use case._

**Comment by peterwoodworth** [comment]
Seems like this issue has been appropriately answered, if you don't have any more questions i think we can close this out. @dfonnegra do you still need help with this?
